src/irmin-pack/IO.ml

```diff
@@ -146,10 +185,32 @@ module Unix : S = struct
     in
     aux dirname (fun () -> ())
 
+  let raw ~mode ~version file =
+    let x = Unix.openfile file Unix.[ O_CREAT; mode; O_CLOEXEC ] 0o644 in
+    let raw = Raw.v x in
+    let header = { Raw.Header.version; offset = 0L; generation = 0L } in
+    Raw.Header.set raw header;
+    raw
+
   let clear t =
-    t.offset <- 0L;
-    t.flushed <- header;
-    Buffer.clear t.buf
+    Log.debug (fun l -> l "clear %s" t.file);
+    if t.readonly then invalid_arg "Read-only IO cannot be cleared";
+    Buffer.clear t.buf;
+    (* no-op if the file is already empty; this is to avoid bumping
+       the version number when this is not necessary. *)
+    if t.offset = 0L then ()
+    else (
+      t.offset <- 0L;
+      t.flushed <- header t;
```

Comment:
I think this line should be _below_ the generation upgrade, since that will change the value of `header t`. (Either way, we should probably add a comment make it clear which is correct.)